### PR TITLE
First draft of design

### PR DESCRIPTION
**Pull request by kc1212**
Requesting for comments

### CODE REVIEW

**Comment by kc1212** [comment]
Just wrote up most of the things we discussed. @aaronang @victor-li please have a look and let me know if there's anything unclear or questionable. If I get an OK from both of you guys then I'll merge it.

File: README.md [hunk 1]
```diff
@@ -1 +1,46 @@
 # Alc the Ripper
+Alc the Ripper (Alc for short) is a *state-of-the-art* cloud-based password cracker.
+The user submits a salt, a PBKDF2 digest, and the length of the password, then Alc will use brute force to find the password.
+
+# Design Requirements
+* Automation - Alc should not require any human intervention when running. The user simply submits a request and Alc will do its job and then report the finding.
+* Elasticity - Alc should handle variable user demands and offer the same level of service (in terms of hashes per second) to all the users.
+* Load Balancing - The jobs should be evenly spread out across all the slaves to achieve maximum performance.
+* Reliability - Jobs should checkpointed and restarted from the checkpoint when failures occur.
```
Comment (aaronang) [inline README.md]: English please :trollface:

**Comment by victor-li** [comment]
@kc1212 LGTM

File: README.md [hunk 2]
```diff
@@ -1 +1,46 @@
...
+* Monitoring - Maintain metrics about the whole system to monitor job status, resource usage and so on.
+* Scheduling - TODO
+* Multi-tenancy - TODO
+* Security - TODO
+
+# Design Overview
+Alc is designed to run on IaaS providers such as AWS.
+It uses the master-slave model. Where the master assign tasks to the slaves and the slaves periodically send heartbeat messages and status updates back to the master.
```
Comment (aaronang) [inline README.md]: **, where** the master assign**s**

File: README.md [hunk 3]
```diff
@@ -1 +1,46 @@
...
+
+We achieve elasticity using the [PID controller](https://en.wikipedia.org/wiki/PID_controller).
+Before getting into its functionality, we first define the meaning of resources in Alc.
+The number of resources is the number of tasks that can be carried out in parallel in a time instant, somewhat analogous to [DOP](https://en.wikipedia.org/wiki/Degree_of_parallelism).
+For example, if an EC2 instance can run two tasks in parallel, then that instance represents two units of resource.
+For the PID controller, we define the error as `Ra - Rr`, where `Ra` is the total available resource, and `Rr` is the total required resource.
```
Comment (aaronang) [inline README.md]: Total available resource is equal to how much we can pay AWS. Therefore, I think "total available" is not correctly phrased.

File: README.md [hunk 4]
```diff
@@ -1 +1,46 @@
...
+`Rr` depends on the number of submitted tasks and `Ra` is total number of instances that Alc provisioned multiplied by the DOP of every instance.
+If the number of submitted tasks exceeds the maximum possible resource, e.g. 20 instances each with DOP of 2,
+then the tasks need to be queued and the total required resource is capped at 20 * 2.
```
Comment (aaronang) [inline README.md]: total required ➡️ total available

File: README.md [hunk 5]
```diff
@@ -1 +1,46 @@
...
+The queuing happen on the master.
```
Comment (aaronang) [inline README.md]: happen ➡️ happens

File: README.md [hunk 6]
```diff
@@ -1 +1,46 @@
...
+Since the SHA2 computation is deterministic, it is also possible to estimate the finish time of tasks and push new tasks to the slaves just before the tasks are finished to minimise waiting time on the slave.
+The greedy algorithm is not optimal under normal circumstances, e.g. it does a bad job if all the tasks are short but the final task is long.
+However, in our case it is possible to fix the duration of the tasks by chunking them into sub-tasks and then combining the results after all the sub-tasks are completed.
+For example a task might be - brute force a 6 character lowercase only password, then it is possible to split it into two sub-tasks where the first is responsible to compute all hashes beginning with a-m and the second is responsible for n-z.
+If the duration of all sub-tasks are all the same or similar, we claim the greedy algorithm is optimal.
+We set a global sub-task size in terms of the number of SHA2 hashes.
+
+To achieve reliability on the master, we introduce a backup master. 
+The master periodically synchronise its internal data structure with the backup, and the backup is responsible to detect when the master is offline and then take over if necessary.
```
Comment (aaronang) [inline README.md]: synchronise ➡️ synchronises
responsible to detect ➡️ responsible for detecting

File: README.md [hunk 7]
```diff
@@ -1 +1,46 @@
...
+It is possible to have multiple backup masters. The master (leader) can be elected using the [Bully algorithm](https://en.wikipedia.org/wiki/Bully_algorithm) where new leaders are automatically elected when the current leader goes offline.
+The reliability on the slaves is achieved via the master.
+The master internally stores the task status and slave status, effectively creating checkpoints for every task.
+If the slave misses too many heartbeat messages then the master can assume it crashed and it will create a new slave by provision a new instance and start the task from the last known checkpoint.
```
Comment (aaronang) [inline README.md]: provision ➡️ provisioning?

**Review by aaronang — COMMENTED** [review]
The general idea is good. I just found some grammar mistakes.

File: README.md (hunk 3 above)
Comment (kc1212) [reply README.md]: Changed to "currently available"

File: README.md (hunk 4 above)
Comment (kc1212) [reply README.md]: this is actually the total required resource, the message i want to get across is we don't consider the jobs in the queue as required resource.

**Comment by kc1212** [comment]
Making some fixes regarding your comments @aaronang

**Comment by aaronang** [comment]
@kc1212 Change to Cong the Ripper :trollface:

**Review by aaronang — APPROVED** [review]
Fix merge conflict and you are allowed to merge 😄

**Review by victor-li — APPROVED** [review]

**Pull request merged**